### PR TITLE
THREESCALE-10499  override requests,limits and replicas when hpa enabled

### PR DESCRIPTION
**Pull request by austincunningham**
## What
merge [THREESCALE-10500](https://issues.redhat.com//browse/THREESCALE-10500) first
issue: [THREESCALE-10499](https://issues.redhat.com//browse/THREESCALE-10499)

### Positive case 
Async mode enabled 
We will need non logical backend redis to enable async mode 
- [ ] you can Run [CRO](https://github.com/integr8ly/cloud-resource-operator) locally to provision 2 on cluster redis 
```bash
make cluster/prepare
make run
REDIS_NAME=redis-storage make cluster/seed/redis
REDIS_NAME=redis-que make cluster/seed/redis
```
The redis will be installed in the cloud-resources-operator ns there should be 2 secrets  `redis-que-sec` and `redis-storage-sec` with the port and the uri for the redis
- [ ] once they are created you can stop the CRO operator running
- [ ] create a key/value secret in the 3scale-test ns called `backend-redis` and add 2 name values with the connection uri and port for the env vars REDIS_QUEUES_URL & REDIS_STORAGE_URL 
image

- [ ] Create a project `oc new-project 3scale-test`
- [ ] `make install`
- [ ] `make run`
- [ ] Create a s3 secret
```bash
kubectl apply -f - <<EOF
---
kind: Secret
apiVersion: v1
metadata:
  name: s3-credentials
data:
  AWS_ACCESS_KEY_ID: UkVQTEFDRV9NRQ==
  AWS_BUCKET: UkVQTEFDRV9NRQ==
  AWS_REGION: UkVQTEFDRV9NRQ==
  AWS_SECRET_ACCESS_KEY: UkVQTEFDRV9NRQ==
type: Opaque
EOF
```
- [ ] get the default application router
```bash
DOMAIN=$(oc get routes console -n openshift-console -o json | jq -r '.status.ingress[0].routerCanonicalHostname' | sed 's/router-default.//')
```
- [ ] Create an apimanager CR using that secret e.g.
```bash
kubectl apply -f - <<EOF
---
apiVersion: apps.3scale.net/v1alpha1
kind: APIManager
metadata:
  name: apimanager-sample
spec:
  system:
    fileStorage:
      simpleStorageService:
        configurationSecretRef:
          name: s3-credentials
  wildcardDomain: $DOMAIN
EOF
```
- update apimanager CR spec:
```yaml
spec:
  backend:
    listenerSpec:
      replicas: 4
      resources:
        limits:
          cpu: 2000m
          memory: 500Mi
        requests:
          cpu: 300m
          memory: 150Mi
    workerSpec:
      replicas: 4
      resources:
        limits:
          cpu: 2000m
          memory: 500Mi
        requests:
          cpu: 300m
          memory: 150Mi
  apicast:
    productionSpec:
      replicas: 4
      resources:
        limits:
          cpu: 2000m
          memory: 500Mi
        requests:
          cpu: 300m
          memory: 150Mi
```
- [ ] Wait for apimanger CR to report completed

- The outcome here is that apimanger pods will start up, with 4 replicas and above limits and request

- [ ] Set the `hpa: true` on 
- `spec.apicast.productionSpec.hpa: true`
- `spec.backend.workerSpec.hpa: true`
- `spec.backend.listenerSpec.hpa: true`
- Outcome will be that the hpa object will be created, with 1-5 pods and initial value of pods matching the previous value which was - 4
- Let it run for about 10 minutes, the number of replicas should reduce over time
- Confirm that a warning condition is visible on the apimanager CR indicating that requests, limits, and replicas values are ignored because HPA took over.
```yaml
status:
  conditions:
    - lastTransitionTime: '2024-01-18T10:00:33Z'
      status: 'True'
      type: Available
    - lastTransitionTime: '2024-01-17T12:22:58Z'
      message: >-
        HorizontalPodAutoscaling (Hpa) enabled overrides values applied to
        request, limits and replicas
      reason: HPA
      status: 'True'
      type: Warning
```
- Confirm that requests and limits are set to the same values  in the apicast-production, backend-listener and backend-worker deploymentConifgs

- Set the HPA to false / or remove it. The resource limits, requests and replicas should go back to old values (or the ones specified in the APICast CR) and the warning message should be removed from the CR.

- [ ] confirm that 3 hpa objects are created with the default spec
```bash
oc get hpa
NAME                 REFERENCE                             TARGETS            MINPODS   MAXPODS   REPLICAS   AGE
apicast-production   DeploymentConfig/apicast-production   54%/90%, 0%/90%    1         5         1          69m
backend-listener     DeploymentConfig/backend-listener     13%/90%, 0%/90%    1         5         1          69m
backend-worker       DeploymentConfig/backend-worker       108%/90%, 0%/90%   1         5         5          69m
```
### Negative case
Async mode disabled for backend worker and listener

- [ ] Create a project `oc new-project 3scale-test2`
- [ ] `make install`
- [ ] `make run`
- [ ] Create a s3 secret
```bash
kubectl apply -f - <<EOF
---
kind: Secret
apiVersion: v1
metadata:
  name: s3-credentials
data:
  AWS_ACCESS_KEY_ID: UkVQTEFDRV9NRQ==
  AWS_BUCKET: UkVQTEFDRV9NRQ==
  AWS_REGION: UkVQTEFDRV9NRQ==
  AWS_SECRET_ACCESS_KEY: UkVQTEFDRV9NRQ==
type: Opaque
EOF
```
- [ ] get the default application router
```bash
DOMAIN=$(oc get routes console -n openshift-console -o json | jq -r '.status.ingress[0].routerCanonicalHostname' | sed 's/router-default.//')
```
- [ ] Create an apimanager CR using that secret e.g.
```bash
kubectl apply -f - <<EOF
---
apiVersion: apps.3scale.net/v1alpha1
kind: APIManager
metadata:
  name: apimanager-sample
spec:
  system:
    fileStorage:
      simpleStorageService:
        configurationSecretRef:
          name: s3-credentials
  wildcardDomain: $DOMAIN
EOF
```
- [ ] once your installation is finished set the following fields in the apmanager cr to true
- `spec.apicast.productionSpec.hpa: true`
- `spec.backend.workerSpec.hpa: true`
- `spec.backend.listenerSpec.hpa: true`

- [ ] confirm that only the hpa objects for apicast-production is created with the default spec
```bash
oc get hpa
NAME                 REFERENCE                             TARGETS            MINPODS   MAXPODS   REPLICAS   AGE
apicast-production   DeploymentConfig/apicast-production   54%/90%, 0%/90%    1         5         1          69m
```
- [ ] Confirm that the status is updated with the following messages
```yaml
status:
  conditions:
    - lastTransitionTime: '2024-01-17T12:29:42Z'
      status: 'False'
      type: Available
    - lastTransitionTime: '2024-01-17T12:36:09Z'
      message: >-
        HorizontalPodAutoscaling (Hpa) Logical Redis databases detected for
        backend, logical Redis database not compatible with async mode, Hpa
        requires async mode for backend, Hpa not enabled for backend
      reason: HPA
      status: 'True'
      type: Configuration Warning
    - lastTransitionTime: '2024-01-17T14:02:26Z'
      message: >-
        HorizontalPodAutoscaling (Hpa) enabled overrides values applied to
        request, limits and replicas
      reason: HPA
      status: 'True'
      type: Warning
```

### CODE REVIEW

File: doc/operator-user-guide.md [hunk 1]
```diff
@@ -793,7 +793,8 @@ spec:
           hpa: true
 ```
 Removing hpa field or setting enabled to false will remove the HPA for the component. 
-Once `enabled: true` you can manually edit the HPA instances that are created to optimize your [configuration](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/). 
+Once `enabled: true` you can manually edit the HPA instances that are created to optimize your [configuration](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/).
```
Comment (MStokluska) [inline doc/operator-user-guide.md:796]: this seems out of date?

File: pkg/3scale/amp/operator/apicast_options_provider.go [hunk 2]
```diff
@@ -30,6 +30,7 @@ const (
 	APIcastEnvironmentCMAnnotation             = "apps.3scale.net/env-configmap-hash"
 	PodPrioritySystemNodeCritical              = "system-node-critical"
 	CustomPoliciesSecretResverAnnotationPrefix = "apimanager.apps.3scale.net/custompolicy-secret-resource-version-"
+	HPAEnabledWarning                          = "HorzontalPodAutoscaling enabled this overrides values applied to request, limits and replicas"
```
Comment (MStokluska) [inline pkg/3scale/amp/operator/apicast_options_provider.go:33]:
```suggestion
	HPAEnabledWarning                          = "HorizontalPodAutoscaling enabled overrides values applied to request, limits and replicas"
```

File: pkg/3scale/amp/operator/backend_reconciler.go [hunk 3]
```diff
@@ -190,3 +183,14 @@ func Backend(apimanager *appsv1alpha1.APIManager, client client.Client) (*compon
 	}
 	return component.NewBackend(opts), nil
 }
+
+func GetBackendRedisSecret(apimanagerNs string, ctx context.Context, client2 client.Client) (string, string) {
```
Comment (MStokluska) [inline pkg/3scale/amp/operator/backend_reconciler.go:187]: Just out of curiosity, any reason for calling the client param "client2"?

File: controllers/apps/apimanager_status_reconciler.go [hunk 4]
```diff
@@ -224,3 +226,47 @@ func (s *APIManagerStatusReconciler) defaultRoutesReady() (bool, error) {
 
 	return allDefaultRoutesReady, nil
 }
+
+func (s *APIManagerStatusReconciler) reconcileHpaWarningMessages(conditions *common.Conditions, cr *appsv1alpha1.APIManager, client2 client.Client) {
```
Comment (MStokluska) [inline controllers/apps/apimanager_status_reconciler.go:230]: we can use s.Client() here instead.

File: pkg/3scale/amp/operator/apicast_options_provider.go (hunk 2 above)
Comment (austincunningham) [reply pkg/3scale/amp/operator/apicast_options_provider.go:33]: Looks like I am not using this const, will remove it

File: doc/operator-user-guide.md (hunk 1 above)
Comment (austincunningham) [reply doc/operator-user-guide.md:796]: yep should be `hpa: true`

File: controllers/apps/apimanager_status_reconciler.go (hunk 4 above)
Comment (MStokluska) [reply controllers/apps/apimanager_status_reconciler.go:230]: @austincunningham we don't use client2 now, can we remove it?

Comment (austincunningham) [reply controllers/apps/apimanager_status_reconciler.go]: good point

**Comment by MStokluska** [comment]
/lgtm
Thanks Austin

**Pull request merged**